h2/src/main/org/h2/mvstore/db/TransactionStore.java

```diff
@@ -1309,8 +1309,11 @@ public void clear() {
          * @return the first key, or null if empty
          */
         public K firstKey() {
-            Iterator<K> it = keyIterator(null);
-            return it.hasNext() ? it.next() : null;
+            K k = map.firstKey();
+            while (k != null && get(k) == null) {
+                k = map.higherKey(k);
+            }
+            return k;
         }
```

Comment:
While your are trying to eliminate object creation here, performance gains are not obvious to me. Proposed implementation does root-to-leaf traversal at least twice, and then twice for each new uncommitted entry encountered (unlikely event, of course, but Murphy's law proves itself times and times again), while iterator does single root-to-leaf traversal and then navigates within a leaf.

Reply:
Iterator with `includeUncommitted = false` calls get() by itself anyway for each found match.

Reply:
Yes, it does, but it does not have to be this way - iterator can be changed to avoid traversal, because it has the value right at it's fingertips.
If interested, please [take a look here](https://github.com/andreitokar/h2database/tree/compact). This is my attempt to make MVStore operations non-blocking among other things.

Reply:
If iterators are going to be optimized, I can send another pull request that just adds `ceilingKey()`, `floorKey()` (I need them), normalizes loops, but doesn't replaces iterators with these methods.